doc/developer/design/20230528_managed_clusters.md

```diff
@@ -0,0 +1,357 @@
+  foo
+  bar
+
+> CREATE CLUSTER c REPLICAS (
+  r1 (SIZE 'small'),
+  r3 (SIZE 'medium')
+);
+> ALTER CLUSTER c SET (MANAGED);
+ERROR: cannot convert cluster "c" to a managed cluster
+DETAIL: The replicas do not match required naming pattern `r1`, `r2`, .... The
+first missing replica is "r2".
+```
+
+#### Converting a managed cluster to an unmanaged cluster
+
+`ALTER CLUSTER` will support disabling the `MANAGED` option for a cluster to
+convert a managed cluster to an unmanaged cluster:
+
+```sql
+ALTER CLUSTER c SET (MANAGED = false);
+```
+
+There are no constraints on the conversion. All managed clusters can be
+converted to unmanaged clusters.
+
+The conversion preserves the existing replicas of the cluster. Users can then
+create or drop replicas of the cluster using `{CREATE|DROP} CLUSTER REPLICA`.
+
+#### `{CREATE|DROP} CLUSTER REPLICA`
+
+The `CREATE CLUSTER REPLICA` and `DROP CLUSTER REPLICA` statements will be
+restricted to operations on replicas of an unmanaged cluster.
+
+Materialize will produce an error message like the following when attempting
+to manually manipulate replicas of a managed cluster:
+
+```
+> CREATE CLUSTER c SIZE 'small';
+> DROP CLUSTER REPLICA c1.r1;
+ERROR: cannot drop replica of managed cluster "c"
+HINT: Use ALTER CLUSTER to change the cluster's size and replication factor, or
+convert the cluster to an unmanaged cluster by running ALTER CLUSTER c SET (MANAGED = false).
+```
+
+#### `ALTER CLUSTER [REPLICA] ... RENAME TO`
+
+`ALTER CLUSTER` and `ALTER CLUSTER REPLICA` will learn to support renaming
+clusters and cluster replicas, respectively:
+
+```sql
+-- Rename cluster c1 to c2.
+ALTER CLUSTER c1 RENAME TO c2;
+
+-- Rename replica r1 of cluster c1 to r2.
+ALTER CLUSTER REPLICA c1.r1 RENAME TO r2;
+```
+
+`ALTER CLUSTER REPLICA` will not permit renaming a replica of a managed cluster.
+
+#### `mz_clusters`
+
+`mz_clusters` will grow three new columns:
+
+Field                | Type               | Meaning
+---------------------|--------------------|--------
+`managed`            | `bool`             | Whether the cluster has automatically managed replicas.
+`size`               | `text`             | If the cluster is managed, the desired size of the cluster's replicas. If the cluster is unmanaged, `NULL`.
+`replication_factor` | `bigint`           | If the cluster is managed, the desired number of replicas of the cluster. If the cluster is unmanaged, `NULL`.
+
+## Alternatives
+
+We could wait to improve the API until we've built [dynamic cluster scheduling].
+
+## Open questions
+
+* Is it acceptable that `ALTER CLUSTER ... SET (SIZE = ...)` causes downtime?
+
+* Should we infer the `MANAGED` option based on whether `SIZE` is set? If so,
+  users would convert to a managed cluster by running
+  `ALTER CLUSTER unmanaged SET (SIZE = ...)`, and convert to an unmanaged
+  cluster by running `ALTER CLUSTER managed RESET (SIZE)`.
+
+  I personally prefer the explictness of
```

Comment:
```suggestion
  I personally prefer the explicitness of
```

Reply:
Since there will be a `managed ` column in `mz_clusters` and we'll have to introduce users to the difference, might as well be explicit.

Reply:
Well in this alternative we’d omit the `managed` column, I think, and let managed be implied by the `size` column being non null.

Reply:
Ah, so it would be similar to what we're doing with linked clusters today? If we expect users to default to managed clusters, then I could live with not being explicit about this either.

Reply:
Yep, exactly, that's the alternatively: try to sweep unmanaged clusters under the rug as hard as possible, just like we've tried to hide linked clusters as much as possible.

I'm less convinced we need to hide unmanaged clusters, though. They are less awful than linked clusters.